File: src/medium/top-frequent-words.ts
export const topFrequentWords = (words: string[], k: number): string[] => {
    const freq = new Map<string, number>();
    words.forEach((word) => freq.set(word, (freq.get(word) ?? 0) + 1));
    return freq
        .entries()
        .toArray()
        .toSorted((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]))
        .slice(0, k)
        .map(([k, v]) => k);
};

File: lib/medium/top_frequent_words.dart
import 'package:collection/collection.dart';

List<String> topFrequentWords(List<String> words, int k) {
  var freq = <String, int>{};
  for (String word in words) {
    freq[word] = (freq[word] ?? 0) + 1;
  }
  var entries = freq.entries.toList();

  entries.sorted((a, b) {
    final compare = b.value.compareTo(a.value);
    if (compare != 0) return compare;
    return a.key.compareTo(b.key);
  });
  return entries.take(k).map((entry) => entry.key).toList();
}
